dask_image/skexposure/histogram_matching.py

```diff
@@ -0,0 +1,97 @@
+import dask
+import dask.array as da
+import numpy as np
+
+# Only these types are allowed, since some intermediate memory storage scales
+# as the number of unique values in the range of the dtype.
+DTYPES = (np.uint8, np.uint16, np.int8, np.int16)
+
+
+def _match_cumulative_cdf(source, template):
+    """
+    Return modified source array so that the cumulative density function of
+    its values matches the cumulative density function of the template.
+    """
+    if ((source.dtype not in DTYPES) or (template.dtype not in DTYPES)):
+        raise ValueError('Parallelization of match_histograms implemented '
+                         'only for small integer types, not: %s %s' %
+                         (source.dtype, template.dtype))
+
+    _, src_unique_indices, src_counts = da.unique(source.ravel(),
+                                                  return_inverse=True,
+                                                  return_counts=True)
+    tmpl_values, tmpl_counts = da.unique(template.ravel(), return_counts=True)
+
+    # calculate normalized quantiles for each array
+    src_quantiles = da.cumsum(src_counts) / source.size
+    tmpl_quantiles = da.cumsum(tmpl_counts) / template.size
+
+    # The interpolation is a bottleneck and must be done on local node.
+    # This requires an in-memory array which could have a length equal in size
+    # to the unique values of the underlying dtype.  This is why we limit to
+    # small integers.
+    interp_a_values = dask.delayed(_interpolate)(src_quantiles, tmpl_quantiles, tmpl_values)
+    interp_a_values = da.from_delayed(interp_a_values, dtype=np.float, shape=src_quantiles.shape)
```

Comment:
Also, there's a `dtype=float` in here, which means the output result is always float type. Ideally, we'd try and preserve the datatype so it matches the inputs.

Reply:
The original code from skimage does not preserve datatype, so I thought it best not to surprise people switching over.

Reply:
The reason I had for suggesting it is that unnecessarily storing 8-bit integer data in a float array takes more memory, so for very large dask arrays it might mean slower performance for whatever you do next with the output of `match_historgrams`. 

I don't know how big of an effect this might be (it probably depends on the specific use case), but it does make sense to see if there's an easy-ish way to avoid that potential problem altogether.

Reply:
I agree with your worry.  I'll add a parameter, which defaults to the skimage behavior.

I pushed it all the way into the `_interpolate` step, though the big win is just in the `map_blocks`.  

It would be amazing if dask optimization would have fused the `astype` from a `match_histograms(...).astype(np.int16)` into the `map_blocks` automatically...